### PR TITLE
Changes to handle the Max payload error

### PR DESCRIPTION
**Pull request by rakhimundhada15**
### Problem Description
For O365 collector when we pulling the data for 1hr duration, if data size > 10Mb then al-collector.js  [buildPayload](https://github.com/alertlogic/al-collector-js/blob/master/al_log.js#L97) method return Max payload issue.

### Solution Description
Reduce the pull interval by 2 and accordingly updated the state value

### CODE REVIEW

File: paws_collector.js [hunk 1]
```diff
@@ -461,6 +439,49 @@ class PawsCollector extends AlAwsCollector {
         });
     };
 
+    batchLogProcess(logs, privCollectorState, nextInvocationTimeout, asyncCallback) {
+        let collector = this;
+        return collector.processLog(logs, collector.pawsFormatLog.bind(collector), null, (err) => {
+           
+            if (err) {
+                // MS api return the logs but if build payload size > 10MB, our buildPayload method return Max payload issue. So we split logs array the in two part and call the processLog again.
```
Comment (kkuzmin) [inline paws_collector.js:447]: please make it a multiline comment

File: paws_collector.js [hunk 2]
```diff
@@ -461,6 +439,49 @@ class PawsCollector extends AlAwsCollector {
...
+               
+                if ((typeof err === 'string' && err.indexOf('Maximum payload size exceeded') !== -1)) {
+                    let batchLogs = [];
+                    let halfLogs = [];
+                    let loglength = logs.length / 2;
+                    for (const index in logs) {
+                        console.log(index,logs[index]);
```
Comment (kkuzmin) [inline paws_collector.js:454]: please remove debug logs.

File: paws_collector.js [hunk 3]
```diff
@@ -461,6 +439,49 @@ class PawsCollector extends AlAwsCollector {
         });
     };
 
+    batchLogProcess(logs, privCollectorState, nextInvocationTimeout, asyncCallback) {
```
Comment (kkuzmin) [inline paws_collector.js:442]: please rename `asyncCallback ` to `callback`

File: paws_collector.js [hunk 4]
```diff
@@ -461,6 +439,49 @@ class PawsCollector extends AlAwsCollector {
...
+        return collector.processLog(logs, collector.pawsFormatLog.bind(collector), null, (err) => {
+           
+            if (err) {
+                // MS api return the logs but if build payload size > 10MB, our buildPayload method return Max payload issue. So we split logs array the in two part and call the processLog again.
+               
+                if ((typeof err === 'string' && err.indexOf('Maximum payload size exceeded') !== -1)) {
+                    let batchLogs = [];
+                    let halfLogs = [];
+                    let loglength = logs.length / 2;
+                    for (const index in logs) {
```
Comment (kkuzmin) [inline paws_collector.js:453]: it should be easier to use [`array.slice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice) to split an array into halves.

Comment (kkuzmin) [reply paws_collector.js:453]: for example you can have a list of indexes to be handled in a `map`, like:
```
const indexArray = [
    {start: 0, stop: half}, 
    {start:half+1, logs.length}
];

indexArray.map(logsPart => {
    collector.batchLogProcess(logs.slice(logsPart.start, logsPart.stop, ....);
});

```

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: paws_collector.js [hunk 5]
```diff
@@ -447,6 +439,49 @@ class PawsCollector extends AlAwsCollector {
...
+                            halfLogs = [];
+                            loglength = logs.length;
+                            halfLogs.push(logs[index]);
+                            if (parseInt(index) == logs.length-1) {
+                                batchLogs.push(halfLogs);
+                            }
+                        }
+                    }
+                    batchLogs.map(batchLog => {
+                         collector.batchLogProcess(batchLog, privCollectorState, nextInvocationTimeout, asyncCallback);
```
Comment (kkuzmin) [inline paws_collector.js:471]: we can't call `asyncCallback` here, becasue once the first batch is sent the callback will be returned and invocation will end without waiting for the 2nd batch to be sent. 
We need to send batches one by one and probably use `array.reduce()`

File: paws_collector.js (hunk 4 above)
Comment (kkuzmin) [reply paws_collector.js:453]: and use `indexArray.reduce()` instead of `indexArray.map()`

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: paws_collector.js (hunk 4 above)
Comment (rakhimundhada15) [reply paws_collector.js:453]: indexArray.reduce() return the single value.  I am checking  foe Async.reduce().

File: paws_collector.js (hunk 5 above)
Comment (rakhimundhada15) [reply paws_collector.js:471]: Using async.reduce().

File: paws_collector.js [hunk 6]
```diff
@@ -495,6 +487,37 @@ class PawsCollector extends AlAwsCollector {
...
+                 */
+                if ((typeof err === 'string' && err.indexOf('Maximum payload size exceeded') !== -1)) {
+
+                    let half = logs.length / 2;
+                    const indexArray = [
+                        { start: 0, stop: half },
+                        { start: half, stop: logs.length }
+                    ];
+
+                    async.reduce(indexArray, 0, (memo, logpart, asyncCallback) => {
```
Comment (kkuzmin) [inline paws_collector.js:507]: reduce returns a promise if 4th param `callback` is not pass
We need to add a new callback here and check the output of each `batchLogProcess`

Comment (kkuzmin) [reply paws_collector.js:507]: I think we can also use `map()` here, but I don't have strong preference.

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: paws_collector.js [hunk 7]
```diff
@@ -495,6 +487,37 @@ class PawsCollector extends AlAwsCollector {
...
+        return collector.processLog(logs, collector.pawsFormatLog.bind(collector), null,  (err) => {
+           
+            if (err) {
+                /**
+                 * MS api return the logs but if build payload size > 10MB, and split logs array the in two part.
+                 * Call the processLog again to send Logmsgs to ingest. 
+                 */
+                if ((typeof err === 'string' && err.indexOf('Maximum payload size exceeded') !== -1)) {
+
+                    let half = logs.length / 2;
```
Comment (kkuzmin) [inline paws_collector.js:501]: please make it `const`
and also wrap it into `Math.floor()`

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: paws_collector.js [hunk 8]
```diff
@@ -495,6 +487,37 @@ class PawsCollector extends AlAwsCollector {
...
+                    const indexArray = [
+                        { start: 0, stop: half },
+                        { start: half, stop: logs.length }
```
[the remark this answers is not in the document]
Comment (rakhimundhada15) [reply paws_collector.js:504]: array.slice exclude(stop: half)  last index value, so including that value in next half(start: half)

File: paws_collector.js [hunk 9]
```diff
@@ -495,6 +487,48 @@ class PawsCollector extends AlAwsCollector {
...
+                                    reject(err);
+                                }
+                                else {
+                                    resolve(res);
+                                }
+                            });
+                        });
+                    });
+                    const results = await Promise.all(promises);
+                    return callback(null, results[results.length-1], nextInvocationTimeout);
```
Comment (kkuzmin) [inline paws_collector.js:520]: I think we need to return `privCollectorState ` instead of `results[results.length-1]`

Comment (rakhimundhada15) [reply paws_collector.js:520]: both are returning same value. I will use privCollectorState.

**Review by kkuzmin — APPROVED** [review]

**Pull request merged**